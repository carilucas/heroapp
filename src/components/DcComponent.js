import React from 'react'
import { HeroList } from './heros/HeroList'

export const DcComponent = () => {
   return (
      <div>
         <h1>DcComponent</h1>
         <hr/>
         <HeroList publisher='DC Comics' />
      </div>
   )
}
