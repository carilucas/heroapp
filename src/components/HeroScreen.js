import React, { useContext } from 'react'
import { Redirect, useParams } from 'react-router';
import { getHeroById } from '../helpers/getHeroById';
import {AuthCotext} from '../auth/authContext'




export const HeroScreen = ({history}) => {
   const {heroId} = useParams();
   const hero = getHeroById(heroId);

   const {heroImages} = useContext(AuthCotext)
   
   if (!hero) {
      
      return <Redirect to="/" />
   }

   const {id,superhero,publisher,alter_ego,first_appearance} = hero;
   

   const handleReturn = ()=>{
      if (history.length <= 2) {
         history.push('/')
      }else{
         history.goBack();
      }
   }
   return (
      <div>
         <h1>{superhero} </h1>
         <hr/>
         <div className="row">
            <div className="col-sm-4 animate__animated animate__fadeIn">
               <img 
                  // src={`../assets/heroes/${id}.jpg`} 
                  src={ heroImages( `./${id}.jpg` ).default }
                  alt={superhero} 
                  className="img-fluid" 
               />
            </div>
            <div className="col-sm-8 animate__animated animate__fadeIn">
            <ul className="list-group">
               <li className="list-group-item">Publisher: {publisher}</li>
               <li className="list-group-item">Alter Ego: {alter_ego}</li>
               <li className="list-group-item">First appearance: {first_appearance}</li>
               <button
                  className="btn btn-primary mt-3"
                  onClick={ handleReturn }
               >
                  Go back
               </button>
            </ul>
            </div>
         </div>
      </div>
   )
}
