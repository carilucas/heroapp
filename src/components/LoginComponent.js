import React, { useContext } from 'react';
import { AuthCotext } from '../auth/authContext';
import { LOGIN } from '../auth/types';



export const LoginComponent = ({history}) => {

   const { dispatchUser } = useContext(AuthCotext);
   
   const lastPath = localStorage.getItem('lastPath') || '/';

   
   const handleLogin = ()=>{
      dispatchUser({
         type: LOGIN,
         payload:{
            name:'Carlos',
            logged: true
         }
      })
      history.push(lastPath)
   }



   return (
      <div className="container mt-5">
         <h1>LoginComponent</h1>
         <hr/>
         <button 
            className="btn btn-primary"
            onClick={ handleLogin }
         >
            Login
         </button>
      </div>
   )
}
