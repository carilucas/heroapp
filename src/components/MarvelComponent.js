import React from 'react'
import { HeroList } from './heros/HeroList'

export const MarvelComponent = () => {
   return (
      <div>
         <h1>MarvelComponent</h1>
         <hr/>
         <HeroList publisher='Marvel Comics' />
      </div>
   )
}
