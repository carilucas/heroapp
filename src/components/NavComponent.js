import React, { useContext } from 'react'
import { Link, NavLink, useHistory } from 'react-router-dom'
import { AuthCotext } from '../auth/authContext'
import { LOGOUT } from '../auth/types'

export const Navbar = () => {
      const history = useHistory();
      const { user:{name}, dispatchUser } = useContext(AuthCotext);

      
      
      const handleLogout = ()=>{
            history.replace('/login');
            dispatchUser({
                  type: LOGOUT
            });
            
      }

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">         
            <div className="container">
               <Link 
                  className="navbar-brand" 
                  to="/"
               >
                  Hero App
               </Link>

               <div className="navbar-collapse">
                  <div className="navbar-nav">

                     <NavLink 
                           activeClassName="active"
                           className="nav-item nav-link" 
                           exact
                           to="/marvel"
                     >
                           Marvel
                     </NavLink>

                     <NavLink 
                           activeClassName="active"
                           className="nav-item nav-link" 
                           exact
                           to="/dc"
                     >
                           DC
                     </NavLink>
                     
                     <NavLink 
                           activeClassName="active"
                           className="nav-item nav-link" 
                           exact
                           to="/search"
                     >
                           Search 
                     </NavLink>
                  </div>
               </div>

               <div className="navbar-collapse collapse w-100 order-3 dual-collapse2">
                  <ul className="navbar-nav ms-auto">
                        <li>
                              <span className="nav-item nav-link" style={{color:'cyan'}}>{name}</span>
                        </li>
                        <li>
                              <button 
                                    className="nav-item nav-link btn btn-outline-info" 
                                    onClick={ handleLogout }
                              >
                                    Logout
                              </button>
                        </li>
                  </ul>
               </div>
            </div>
        </nav>
    )
}