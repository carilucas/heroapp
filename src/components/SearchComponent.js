import React, { useMemo } from 'react';
import { useForm } from '../hooks/useForm';
import queryString  from 'query-string';
import { getHeroByName } from '../helpers/getHeroByName';
import { HeroListItem } from './heros/HeroListItem';
import { useLocation } from 'react-router';

export const SearchComponent = ({history}) => {
   
   const location = useLocation();
   
   const {hero=''} = queryString.parse( location.search );

   const heros = useMemo(() => getHeroByName(hero), [hero]);

   const [value, handleInputChange ] = useForm({searchText:hero});
   const {searchText} = value; 
   
   const handleSubmit = (e)=>{
      e.preventDefault();

      history.push(`?hero=${searchText}`)
      
   }


   return (
      <div>
         <h1>Search</h1>
         <hr/>
         <div className="row">
            <div className="col-sm-3">
               <form 
                  onSubmit = { handleSubmit }
                  
               >
                  <input 
                     type="text" 
                     className="form-control"
                     name="searchText"
                     onChange = { handleInputChange }
                     autoComplete="off"
                     value={ searchText }
                  />
                  <button
                     className="btn btn-primary mt-3"
                  >
                     Search
                  </button>
               </form>
            </div>
            <div className="col-sm-9">
               <div className="row animate__animated animate__fadeIn">
                  {
                     hero === '' 
                        && <div className="alert alert-info">Enter a hero name</div>
                  }
                  {
                     hero !== '' && heros.length < 1
                        && <div className="alert alert-danger">There is no matches for {hero}</div>
                  }
                  {
                     heros.map( hero => (
                        <HeroListItem {...hero} key={hero.id}/>
                     ))
                  }
               </div>
            </div>
         </div>
      </div>
   )
}
