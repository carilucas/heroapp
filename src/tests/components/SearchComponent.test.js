import { SearchComponent } from "../../components/SearchComponent";
import { mount } from 'enzyme';
import { MemoryRouter, Route } from "react-router";

// const location = jest.fn();

describe('Pruebas en <SearchComponent>', () => {
   const history = {
      push: jest.fn()
   }
   test('Debe mostrar el Snapshot', () => {
      
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search']}>
            <Route path="/search" component={SearchComponent} />
         </MemoryRouter>
      )
      
      expect( wrapper ).toMatchSnapshot();
      expect( wrapper.find('.alert-info').text().trim()).toBe('Enter a hero name')
   });

   test('Debe mostrar batman y el input con el valor del query string', () => {
      
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?hero=batman']}>
            <Route path="/search" component={SearchComponent} />
         </MemoryRouter>
      )

      expect( wrapper ).toMatchSnapshot();
      expect( wrapper.find('input').prop('value') ).toBe('batman');
      expect( wrapper.find('.card').exists() ).toBe(true);
      expect( wrapper.find('.card-title').text().trim() ).toBe('Batman');

   });

   test('Debe mostrar un error si no se encuentra el hero', () => {
      
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?hero=barbero']} >
            <Route path='/search' component={ SearchComponent } />
         </MemoryRouter>
      );

      expect( wrapper.find('input').prop('value')).toBe('barbero');

      expect( wrapper.find('.alert-danger').exists()).toBe(true);

   });

   test('Debe de llamar el push del history', () => {
      const history = {
         push: jest.fn()
      }
      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?hero=batman']} >
            <Route 
               path='/search' 
               component={ () => <SearchComponent history={history}/> }  
            />
         </MemoryRouter>
      );

      wrapper.find('input').simulate('change',{
         target:{
            name:'searchText',
            value: 'batman'
         }
      });

      wrapper.find( 'form' ).prop('onSubmit')({
         preventDefault(){}
      });

      expect( history.push ).toHaveBeenCalledWith( '?hero=batman' );

   })
   
   
   
   

})
