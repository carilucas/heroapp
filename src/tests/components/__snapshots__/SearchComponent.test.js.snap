// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchComponent> Debe mostrar batman y el input con el valor del query string 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-3"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-sm-9"
    >
      <div
        className="row animate__animated animate__fadeIn"
      >
        <div
          className="col-sm-4 my-3"
        >
          <div
            className="card"
          >
            <img
              alt="Batman"
              className="card-img-top"
              src="./assets/heroes/dc-batman.jpg"
            />
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                DC Comics
              </p>
              <button
                className="btn btn-primary"
                onClick={[Function]}
              >
                More Info
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchComponent> Debe mostrar el Snapshot 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-3"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-sm-9"
    >
      <div
        className="row animate__animated animate__fadeIn"
      >
        <div
          className="alert alert-info"
        >
          Enter a hero name
        </div>
      </div>
    </div>
  </div>
</div>
`;
